### ISSUE TITLE
Custom Solidity errors throw Uncaught TypeError with Mocha --parallel

### ISSUE THREAD

**Issue by davidtperk** [NomicFoundation/hardhat#2355]
In Solidity `^0.8.0`, reverting with a custom Solidity error while testing in parallel using Mocha throws the following error:

```
1) Uncaught error outside test suite:
     Uncaught TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'SourceFile'
    |     property 'contracts' -> object with constructor 'Array'
    |     index 0 -> object with constructor 'Contract'
    |     property 'location' -> object with constructor 'SourceLocation'
    --- property 'file' closes the circle
      at stringify (<anonymous>)
      at writeChannelMessage (internal/child_process/serialization.js:120:20)
      at process.target._send (internal/child_process.js:812:17)
      at process.target.send (internal/child_process.js:710:19)
      at /Users/davidperkins/Desktop/hardhat-examples/parallel-tests/node_modules/workerpool/src/worker.js:131:22
```

Removing the `--parallel` flag from `mocha` throws the expected error: `Error: VM Exception while processing transaction: reverted with custom error 'MyCustomError()'`.

To reproduce, clone my [fork](https://github.com/davidtperk/hardhat-examples) of hardhat-examples. In the parallel-tests folder, run `yarn` followed by `yarn test:parallel`, once with `--parallel` in the npm script and once without.

**Comment by feuGeneA** [comment]
See https://github.com/nomiclabs/hardhat/issues/2331

**Comment by davidtperk** [comment]
@fvictorio's [parallel-tests](https://github.com/fvictorio/hardhat-examples/tree/master/parallel-tests) already runs successfully with `mocha--parallel` and `"mocha": "^8.3.2"`. Is `--parallel` just not "officially" supported?

**Comment by fvictorio** [comment]
@davidtperk: notice that that example uses a somewhat old version of Hardhat. Are you using the latest one? If so, can you please create a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)?

**Comment by davidtperk** [comment]
@fvictorio yes, using `"hardhat": "^2.8.3"`. Here's a reproducible example: https://github.com/davidtperk/hardhat-examples/tree/master/parallel-tests. Run `yarn test:parallel`, once with `--parallel` in the npm script and once without.

**Comment by davidtperk** [comment]
Should be a Hardhat bug because I'm having the same issue in my own project.

**Comment by fvictorio** [comment]
Thanks a lot for finding and reporting this @davidtperk. We were about to release parallel tests support in core, but this is definitely a blocker.

The underlying reason is [this issue](https://github.com/mochajs/mocha/issues/4552). The tl;dr is that our errors have circular references, and parallel mode serializes test errors in a way that doesn't support circular references.

**Comment by alcuadrado** [comment]
This has been fixed in v2.9.0.
